File: src/types/collection.ts
// ---------------------------------------------------------------------------
// v2 Domain Types — Direct mirrors of Rust `domain/` structs
// ---------------------------------------------------------------------------

import type { WorkspacePathRewrite } from './workspace';

export type CollectionKind = 'named' | 'undo_snapshot' | 'unsaved';
export type MemberKind = 'mod' | 'nested' | 'object' | 'root';

export type SwitchResult = {
  success: boolean;
  active_safe: boolean;
  mods_disabled: number;
  mods_restored: number;
  new_signature: string;
  warnings: string[];
  restored_collection_id?: string | null;
};

export type CorridorSnapshot = {
  game_id: string;
  is_safe: boolean;
  signature?: string;
  mod_count?: number;
  timestamp?: string;
  active_collection_id: string | null;
  active_collection_name: string | null;
  active_collection_is_unsaved: boolean;
  is_dirty: boolean;
  undo_collection_id: string | null;
  current_signature: string;
  last_switched_at?: string | null;
  created_at?: string | null;
  current_mods: CollectionMod[];
  current_objects: CollectionObject[];
  current_tree_nodes: PreviewTreeNode[];
  projected_state: ProjectedCollectionState;
};

export type CorridorSwitchPreview = {
  leaving_state_name: string | null;
  leaving_state_is_unsaved: boolean;
  leaving_state_is_safe: boolean;
  leaving_mods: CollectionMod[];
  leaving_objects: CollectionObject[];
  leaving_tree_nodes: PreviewTreeNode[];
  leaving_projected_state: ProjectedCollectionState;
  target_state_name: string | null;
  target_state_is_unsaved: boolean;
  target_state_is_safe: boolean;
  target_state_kind: string;
  target_mods: CollectionMod[];
  target_objects: CollectionObject[];
  target_tree_nodes: PreviewTreeNode[];
  target_projected_state: ProjectedCollectionState;
};

export type CollectionSummary = {
  id: string;
  name: string;
  is_safe: boolean;
  is_unsaved: boolean;
  is_active: boolean;
  is_undo_target: boolean;
  signature: string | null;
  updated_at: string;
  member_count: number;
  mod_count: number;
  game_id?: string;
  created_at?: number;
};

export type CollectionMod = (
  | {
      kind: 'mod';
    }
  | {
      kind: 'nested';
    }
) & {
  collection_id: string;
  mod_id: string | null;
  mod_path: string;
  path_key: string | null;
  mod_path_key: string | null;
  object_id: string;
  display_name: string;
  preview_path: string | null;
  node_type: string | null;
  warnings: string[];
  is_enabled: boolean;
};

export type CollectionObject = {
  kind: 'object';
  collection_id: string;
  object_id: string;
  is_enabled: boolean;
  display_name: string;
  path_key: string;
};

export type CollectionRoot = {
  kind: 'root';
  collection_id: string;
  root_path: string;
  root_path_key: string;
  display_name: string;
  display_name_key: string;
  object_id: string | null;
  object_name: string | null;
  object_type: string | null;
  root_kind: string;
  is_safe: boolean;
  is_enabled: boolean;
  thumbnail_hint: string | null;
  corridor_source: string | null;
  path_key: string;
};

export type CollectionMember = CollectionMod | CollectionObject | CollectionRoot;

export type PreviewTreeNodeKind = 'object' | 'folder' | 'mod';

export type PreviewTreeNode = {
  kind: PreviewTreeNodeKind;
  id: string;
  name: string;
  path: string | null;
  object_id: string | null;
  node_type: string | null;
  is_enabled: boolean;
  is_effectively_active: boolean;
  inactive_reason: string | null;
  show_inactive_chip: boolean;
  status_kind: string | null;
  collapse_children: boolean;
  warnings: string[];
  mod_count: number | null;
  children: PreviewTreeNode[];
};

export type CollectionPreview = {
  collection: CollectionSummary;
  members: CollectionMember[];
  mods: CollectionMod[];
  objects: CollectionObject[];
  roots: CollectionRoot[];
  tree_nodes: PreviewTreeNode[];
  projected_state: ProjectedCollectionState;
};

export type ApplyPreview = {
  collection_name: string;
  current_snapshot: string | null;
  current_mods: CollectionMod[];
  current_objects: CollectionObject[];
  current_tree_nodes: PreviewTreeNode[];
  target_mods: CollectionMod[];
  target_objects: CollectionObject[];
  target_tree_nodes: PreviewTreeNode[];
  current_state_name: string | null;
  current_state_is_unsaved: boolean;
  current_projected_state: ProjectedCollectionState;
  target_projected_state: ProjectedCollectionState;
};

export type ApplyResult = {
  success: boolean;
  mods_enabled: number;
  mods_disabled: number;
  objects_toggled: number;
  undo_collection_id: string | null;
  new_signature: string;
  warnings: string[];
  final_state_name: string | null;
  final_mode: string | null;
  partial_apply: boolean;
  skipped_missing_paths: string[];
  final_state_is_dirty: boolean;
  runtime_path_rewrites: WorkspacePathRewrite[];
};

export type CollectionPathRewrite = {
  from: string;
  to: string;
};

export type CollectionReferenceImpact = {
  affected_collection_count: number;
  affected_collection_names: string[];
  rewritten_paths: CollectionPathRewrite[];
  missing_paths: string[];
};

export type ProjectedObjectState = {
  object_id: string;
  display_name: string;
  path_key: string;
  is_enabled: boolean;
  active_root_count: number;
};

export type ProjectedActiveRoot = {
  object_id: string;
  root_key: string;
  display_name: string;
  root_type: string;
  source_path: string;
  thumbnail_hint: string | null;
  warnings: string[];
  is_missing: boolean;
};

export type ProjectedStateSummary = {
  object_count: number;
  enabled_object_count: number;
  active_root_count: number;
  missing_root_count: number;
};

export type ProjectedCollectionState = {
  object_states: ProjectedObjectState[];
  active_roots: ProjectedActiveRoot[];
  summary: ProjectedStateSummary;
};

export type ApplyProgressSnapshot = {
  game_id: string;
  is_safe: boolean;
  phase: string;
  completed: number;
  total: number;
  current_item: string | null;
  warnings: string[];
  final_state_name: string | null;
  final_mode: string | null;
  success: boolean;
};

export type PinStatus = {
  has_pin: boolean;
  is_locked: boolean;
  attempts_remaining: number;
  lockout_seconds_remaining: number;
};

File: src-tauri/src/domain/collection.rs
use serde::{Deserialize, Serialize};

use crate::domain::workspace::WorkspacePathRewrite;

// ---------------------------------------------------------------------------
// Collection — A named loadout snapshot
// ---------------------------------------------------------------------------

/// The kind of collection. Replaces the `is_last_unsaved` boolean flag.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type, specta::Type)]
#[sqlx(rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum CollectionKind {
    Named,
    UndoSnapshot,
    Unsaved,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, specta::Type)]
#[serde(rename_all = "snake_case")]
pub enum PreviewTreeNodeKind {
    Object,
    Folder,
    Mod,
}

impl CollectionKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Named => "named",
            Self::UndoSnapshot => "undo_snapshot",
            Self::Unsaved => "unsaved",
        }
    }

    pub fn from_db_value(s: &str) -> Self {
        match s {
            "undo_snapshot" => Self::UndoSnapshot,
            "unsaved" => Self::Unsaved,
            _ => Self::Named,
        }
    }
}

/// The kind of member in a collection.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type, specta::Type)]
#[sqlx(rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum MemberKind {
    Mod,
    Nested,
    Object,
    Root,
}

impl MemberKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Mod => "mod",
            Self::Nested => "nested",
            Self::Object => "object",
            Self::Root => "root",
        }
    }

    pub fn from_db_value(s: &str) -> Self {
        match s {
            "nested" => Self::Nested,
            "object" => Self::Object,
            "root" => Self::Root,
            _ => Self::Mod,
        }
    }
}

/// Full collection row from the `collections` table.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct Collection {
    pub id: String,
    pub game_id: String,
    pub name: String,
    pub name_key: String,
    pub is_safe: bool,
    pub is_unsaved: bool,
    pub is_last_unsaved: bool,
    pub last_active: bool,
    pub snapshot_json: Option<String>,
    pub signature: Option<String>,
    pub root_count: i32,
    pub display_mod_count: i32,
    pub member_count: Option<i32>,
    pub created_at: String,
    pub updated_at: String,
}

/// Summary returned in list views.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct CollectionSummary {
    pub id: String,
    pub name: String,
    pub is_safe: bool,
    pub is_unsaved: bool,
    pub is_active: bool,      // Derived from corridor_state
    pub is_undo_target: bool, // Derived from corridor_state
    pub signature: Option<String>,
    pub updated_at: String,
    pub raw_member_count: i32,
    pub mod_count: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct ProjectedObjectState {
    pub object_id: String,
    pub display_name: String,
    pub path_key: String,
    pub is_enabled: bool,
    #[specta(type = f64)]
    pub active_root_count: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct ProjectedActiveRoot {
    pub object_id: String,
    pub root_key: String,
    pub display_name: String,
    pub root_type: String,
    pub source_path: String,
    pub thumbnail_hint: Option<String>,
    pub warnings: Vec<String>,
    pub is_missing: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct ProjectedStateSummary {
    #[specta(type = f64)]
    pub object_count: usize,
    #[specta(type = f64)]
    pub enabled_object_count: usize,
    #[specta(type = f64)]
    pub active_root_count: usize,
    #[specta(type = f64)]
    pub missing_root_count: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct ProjectedCollectionState {
    pub object_states: Vec<ProjectedObjectState>,
    pub active_roots: Vec<ProjectedActiveRoot>,
    pub summary: ProjectedStateSummary,
}

/// A single mod member of a collection (from `collection_mods`).
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct CollectionMod {
    pub kind: MemberKind,
    pub collection_id: String,
    pub mod_id: Option<String>,
    pub mod_path: String,
    pub mod_path_key: Option<String>,
    pub object_id: String,
    pub display_name: Option<String>,
    pub preview_path: Option<String>,
    pub node_type: Option<String>,
    pub warnings: Vec<String>,
    pub is_enabled: bool,
}

/// A single object member of a collection (from `collection_objects`).
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow, specta::Type)]
pub struct CollectionObject {
    pub kind: MemberKind,
    pub collection_id: String,
    pub object_id: String,
    pub is_enabled: bool,
    pub display_name: Option<String>,
    pub path_key: Option<String>,
}

/// A root entry (from `collection_roots`).
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow, specta::Type)]
pub struct CollectionRoot {
    pub kind: MemberKind,
    pub collection_id: String,
    pub root_path: String,
    pub root_path_key: String,
    pub display_name: String,
    pub display_name_key: String,
    pub object_id: Option<String>,
    pub object_name: Option<String>,
    pub object_type: Option<String>,
    pub root_kind: String,
    pub is_safe: bool,
    pub is_enabled: bool,
    pub thumbnail_hint: Option<String>,
    pub corridor_source: Option<String>,
}

/// A unified member type for collections.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum CollectionMember {
    Mod(CollectionMod),
    Object(CollectionObject),
    Root(CollectionRoot),
}

#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct PreviewTreeNode {
    pub kind: PreviewTreeNodeKind,
    pub id: String,
    pub name: String,
    pub path: Option<String>,
    pub object_id: Option<String>,
    pub node_type: Option<String>,
    pub is_enabled: bool,
    pub is_effectively_active: bool,
    pub inactive_reason: Option<String>,
    pub show_inactive_chip: bool,
    pub status_kind: Option<String>,
    pub collapse_children: bool,
    pub warnings: Vec<String>,
    #[specta(type = Option<f64>)]
    pub mod_count: Option<usize>,
    pub children: Vec<PreviewTreeNode>,
}

/// Preview data for a collection.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct CollectionPreview {
    pub collection: CollectionSummary,
    pub members: Vec<CollectionMember>,
    pub mods: Vec<CollectionMod>,
    pub objects: Vec<CollectionObject>,
    pub roots: Vec<CollectionRoot>,
    pub tree_nodes: Vec<PreviewTreeNode>,
    pub projected_state: ProjectedCollectionState,
}

/// Result of applying a collection.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct ApplyResult {
    pub success: bool,
    #[specta(type = f64)]
    pub mods_enabled: usize,
    #[specta(type = f64)]
    pub mods_disabled: usize,
    #[specta(type = f64)]
    pub objects_toggled: usize,
    pub undo_collection_id: Option<String>,
    pub new_signature: String,
    pub warnings: Vec<String>,
    pub final_state_name: Option<String>,
    pub final_mode: Option<String>,
    pub partial_apply: bool,
    pub skipped_missing_paths: Vec<String>,
    pub final_state_is_dirty: bool,
    pub runtime_path_rewrites: Vec<WorkspacePathRewrite>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, specta::Type)]
pub struct CollectionPathRewrite {
    pub from: String,
    pub to: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, specta::Type)]
pub struct CollectionReferenceImpact {
    #[specta(type = f64)]
    pub affected_collection_count: usize,
    pub affected_collection_names: Vec<String>,
    pub rewritten_paths: Vec<CollectionPathRewrite>,
    pub missing_paths: Vec<String>,
}

impl CollectionReferenceImpact {
    pub fn merge(&mut self, next: Self) {
        for name in next.affected_collection_names {
            if !self
                .affected_collection_names
                .iter()
                .any(|existing| existing == &name)
            {
                self.affected_collection_names.push(name);
            }
        }
        for path in next.missing_paths {
            if !self.missing_paths.iter().any(|existing| existing == &path) {
                self.missing_paths.push(path);
            }
        }
        self.rewritten_paths.extend(next.rewritten_paths);
        self.affected_collection_count = self.affected_collection_names.len();
    }
}

/// Input for creating a new collection.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, specta::Type)]
#[serde(rename_all = "snake_case")]
pub enum CreateCollectionMode {
    SaveCurrentState,
    CloneSnapshot,
}

#[derive(Debug, Clone, Deserialize, specta::Type)]
pub struct CreateCollectionInput {
    pub game_id: String,
    pub name: String,
    pub is_safe: bool,
    pub save_mode: Option<CreateCollectionMode>,
    pub source_collection_id: Option<String>,
}

/// Input for updating an existing collection.
#[derive(Debug, Clone, Deserialize, specta::Type)]
pub struct UpdateCollectionInput {
    pub id: String,
    pub game_id: String,
    pub name: Option<String>,
}

/// Preview data for applying a collection (before → after).
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct ApplyPreview {
    pub collection_name: String,
    pub current_snapshot: Option<String>,
    pub current_mods: Vec<CollectionMod>,
    pub current_objects: Vec<CollectionObject>,
    pub current_tree_nodes: Vec<PreviewTreeNode>,
    pub target_mods: Vec<CollectionMod>,
    pub target_objects: Vec<CollectionObject>,
    pub target_tree_nodes: Vec<PreviewTreeNode>,
    pub current_state_name: Option<String>,
    pub current_state_is_unsaved: bool,
    pub current_projected_state: ProjectedCollectionState,
    pub target_projected_state: ProjectedCollectionState,
}

#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct ApplyProgressSnapshot {
    pub game_id: String,
    pub is_safe: bool,
    pub phase: String,
    #[specta(type = f64)]
    pub completed: usize,
    #[specta(type = f64)]
    pub total: usize,
    pub current_item: Option<String>,
    pub warnings: Vec<String>,
    pub final_state_name: Option<String>,
    pub final_mode: Option<String>,
    pub success: bool,
}
